pghoard/restore.py

```diff
@@ -788,8 +807,14 @@ def _write_status_output_to_file(self, output_file):
             }
         )
 
-    def _wait_for_jobs_to_complete(self):
+    def _wait_for_jobs_to_complete(self, pool):
         while self.jobs_in_progress():
+            with self.lock:
+                for item in self.data_files:
+                    if item.id in self.jobs_to_retry:
+                        self.pending_jobs.add(item.id)
+                        self.jobs_to_retry.remove(item.id)
+                        self._queue_job(pool, item)
```

Comment:
Looking at the `_queue_job` implementation it should not matter in practice but in the other place that queues jobs the lock is not being held while calling this function.

Reply:
We need to hold the lock to protect concurrent access to pending_jobs and jobs_to_retry though ?
Since we're looping over jobs_to_retry we can't release the lock for each item.

Reply:
Yea I was thinking of collecting list of items to queue after lock is released. I'm not sure if it's worth the small extra complexity but then again I also haven't checked what the multiprocessing queue does when you queue something, if it does some IPC then it would be better not to hold the lock.

Reply:
Ok, I implemented that.